modules/InfypowerDCSupply/main/power_supply_DCImpl.cpp

```diff
@@ -0,0 +1,176 @@
+               << std::setprecision(1) << "I: " << vc.current_A << " A";
+
+        if (this->last_vc_dbgmsg != dbgmsg.str())
+            EVLOG_debug << dbgmsg.str();
+
+        this->last_vc_dbgmsg = dbgmsg.str();
+    });
+}
+
+void power_supply_DCImpl::handle_setMode(types::power_supply_DC::Mode& mode,
+                                         types::power_supply_DC::ChargingPhase& phase) {
+    std::scoped_lock lock(this->mutex);
+
+    EVLOG_info << "handle_setMode(" << mode << ", " << phase << ")";
+
+    try {
+        switch (mode) {
+        case types::power_supply_DC::Mode::Export:
+            this->mod->controller.set_import_mode(false);
+            this->mod->controller.set_enable(true);
+            break;
+        case types::power_supply_DC::Mode::Import:
+            this->mod->controller.set_import_mode(true);
+            this->mod->controller.set_enable(true);
+            break;
+        default:
+            this->mod->controller.set_enable(false);
+        }
+
+        // finally report the current mode
+        this->publish_mode(mode);
+
+    } catch (std::runtime_error& e) {
+        this->raise_error("power_supply_DC/VendorError", "", e.what());
+    }
+}
+
+void power_supply_DCImpl::handle_setExportVoltageCurrent(double& voltage, double& current) {
+    std::scoped_lock lock(this->mutex);
+
+    EVLOG_info << std::fixed << std::setprecision(1) << "handle_setExportVoltageCurrent(" << voltage << " V, "
+               << current << " A)";
+    try {
+        this->mod->controller.set_voltage_current(voltage, current);
+
+    } catch (std::runtime_error& e) {
+        this->raise_error("power_supply_DC/VendorError", "", e.what());
+    }
+}
+
+void power_supply_DCImpl::handle_setImportVoltageCurrent(double& voltage, double& current) {
+    std::scoped_lock lock(this->mutex);
+
+    EVLOG_debug << std::fixed << std::setprecision(1) << "handle_setImportVoltageCurrent(" << voltage << " V, "
+                << current << " A)";
+    try {
+        // since we do not have any other voltage from EVerest, we assume that
+        // the voltage is carefully chosen and represents the cut-off voltage
+        this->mod->controller.set_import_cutoff_voltage(voltage);
+
+        // the voltage here is probably not important and ignored by the PM
+        // but we re-use the function for setting the current
+        this->mod->controller.set_voltage_current(voltage, current);
+
+    } catch (std::runtime_error& e) {
+        this->raise_error("power_supply_DC/VendorError", "", e.what());
+    }
+}
+
+void power_supply_DCImpl::raise_comm_error(const std::string& errmsg) {
+    bool old_value = this->comm_err_reported.exchange(true);
+
+    // only if the old value (still) was false, then we can raise the error
+    if (!old_value) {
+        const auto err = this->error_factory->create_error("power_supply_DC/CommunicationFault", "", errmsg);
+        this->raise_error(err);
```

Comment:
You introduced a function overload for this. 😉
```suggestion
        this->raise_error("power_supply_DC/CommunicationFault", "", errmsg);
```